### DISCUSSION TITLE
Significant drop in tokenization performance from spaCy 2 to spaCy 3.

### DISCUSSION THREAD

**Discussion by futuremojo** [explosion/spaCy#7519] [Help: Coding & Implementations] [answered]
**Environment**: Google Colab Cloud free tier with NO TPU nor GPU. Just a plain notebook.
**Dataset**: 9,000 articles from [20 Newsgroups](https://scikit-learn.org/0.19/datasets/twenty_newsgroups.html).
<br>
**Task**: Transform articles into TF-IDF vectors using the TfidfVectorizer class from sklearn  and spaCy for preprocessing.

When using spaCy *2* with tagging, ner, and parsing disabled, it takes about 35 seconds to get through 9,000 articles. The code:
```
import spacy
nlp = spacy.load('en_core_web_sm')
unwanted_pipes = ["tagger", "ner", "parser"]

def spacy_tokenizer(doc):
  with nlp.disable_pipes(*unwanted_pipes):
    return [t.text for t in nlp(doc) if \
            not t.is_punct and \
            not t.is_space and \
            t.is_alpha]

%%time
vectorizer = TfidfVectorizer(tokenizer=spacy_tokenizer)
train_feature_vects = vectorizer.fit_transform(train_data)
```
Results using spaCy 2:
CPU times: user 34.1 s, sys: 174 ms, total: 34.3 s
Wall time: 34.6 s

When using spaCy *3* with lemmatizer, tagging, ner, and parsing disabled, it takes over three minutes. The code and filters are identical.
```
import spacy
nlp = spacy.load('en_core_web_sm')
unwanted_pipes = ["lemmatizer", "tagger", "ner", "parser"]

def spacy_tokenizer(doc):
  with nlp.disable_pipes(*unwanted_pipes):
    return [t.text for t in nlp(doc) if \
            not t.is_punct and \
            not t.is_space and \
            t.is_alpha]

%%time
vectorizer = TfidfVectorizer(tokenizer=spacy_tokenizer)
train_feature_vects = vectorizer.fit_transform(train_data)
```
Results using spaCy 3:
CPU times: user 3min 8s, sys: 2.79 s, total: 3min 11s
Wall time: 3min 12s

Both tests used Numpy 1.19.5.

I've tried all sorts of variations such as using TPU, GPU, and nlp.pipe to process in batches but the performance discrepancy still holds.

Any idea what's going on?

**Comment by adrianeboyd** [comment, marked as answer]
If you only want to tokenize, it's easier to use `nlp = spacy.blank("en")` (equivalent to `English()`) rather than disabling components in `en_core_web_sm`. The problem is that there are a few components that are still enabled in your example (`tok2vec`, `attribute_ruler`) and it's probably `tok2vec` that's responsible for most of the speed difference you're seeing. You can always check `nlp.pipe_names` to see which components are currently enabled:

```python
import spacy
nlp = spacy.load("en_core_web_sm")
with nlp.disable_pipes(["lemmatizer", "tagger", "ner", "parser"]):
    print(nlp.pipe_names) # ['tok2vec', 'attribute_ruler']
```

In terms of just the tokenizer, it wouldn't be surprising if the default English tokenizer is ~10% slower in v3. In exchange it handles exceptions with punctuation more consistently and can also handle exceptions that include spaces.

The speed difference mainly depends on whether the tokenizer exceptions (`nlp.tokenizer.rules`) contain any affixes from the prefix/suffix/infix patterns. If your exceptions don't contain any affixes, then the speed should be the same as v2. English has a number of exceptions related to contractions like `don't`, so it's a little slower than in v2.

This timing test is probably a bit out-of-date, but you can see a comparison for a range of languages in this comment: https://github.com/explosion/spaCy/pull/4259#issuecomment-538651788

**Reply by futuremojo** [reply]
Yep, that was it. Thanks, Adriane.

With `spacy.blank("en")`, it took the same amount of time as spaCy 2 with the `parser`, `tagger`, and `ner` components off.

When I needed lemmatization, both versions with no pipeline components disabled took roughly the same time.

As an aside for anyone reading, if you need to retrieve lemmas using spaCy 3, you need the `tok2vec` and `attribute_ruler` pipeline components in addition to `lemmatizer` and `parser`.
